http/http_client.h

```diff
@@ -0,0 +1,432 @@
+        noexcept
+        : verb_(verb), url_(url), callbacks_(callbacks),
+          content_(content), headers_(headers),
+          timeout_(timeout)
+    {
+    }
+
+    void clear()
+    {
+        verb_ = "";
+        url_ = "";
+        callbacks_ = nullptr;
+        content_ = Content();
+        headers_ = RestParams();
+        timeout_ = -1;
+    }
+
+    operator bool ()
+        const
+    {
+        return !verb_.empty();
+    }
+
+    std::string verb_;
+    std::string url_;
+    std::shared_ptr<HttpClientCallbacks> callbacks_;
+    Content content_;
+    RestParams headers_;
+    int timeout_;
+};
+
+
+/****************************************************************************/
+/* HTTP CLIENT IMPL                                                         */
+/****************************************************************************/
+
+struct HttpClientImpl : public AsyncEventSource {
+    HttpClientImpl(const std::string & baseUrl,
+                   int numParallel = 1024, int queueSize = 0)
+        : AsyncEventSource()
+    {
+    }
+
+    HttpClientImpl(HttpClientImpl && other) = default;
+
+    virtual ~HttpClientImpl()
+    {}
+
+    /** Enable debugging */
+    virtual void enableDebug(bool value) = 0;
+
+    /** SSL checks */
+    virtual void enableSSLChecks(bool value) = 0;
+
+    /** Enable the TCP_NODELAY option, also known as the Nagle's algorithm */
+    virtual void enableTcpNoDelay(bool value) = 0;
+
+    /** Use with servers that support HTTP pipelining */
+    virtual void enablePipelining(bool value) = 0;
+
+    /** Enqueue (or perform) the specified request */
+    virtual bool enqueueRequest(const std::string & verb,
+                                const std::string & resource,
+                                const std::shared_ptr<HttpClientCallbacks> & callbacks,
+                                const HttpRequest::Content & content,
+                                const RestParams & queryParams,
+                                const RestParams & headers,
+                                int timeout = -1) = 0;
+
+    /* Returns the number of requests in the queue */
+    virtual size_t queuedRequests() const = 0;
+};
+
+
+/****************************************************************************/
+/* HTTP CLIENT ERROR                                                        */
+/****************************************************************************/
+
+enum struct HttpClientError {
+    None,
+    Unknown,
+    Timeout,
+    HostNotFound,
+    CouldNotConnect,
+    SendError,
+    RecvError
+};
+
+std::ostream & operator << (std::ostream & stream, HttpClientError error);
+
+
+/****************************************************************************/
+/* HTTP CLIENT                                                              */
+/****************************************************************************/
+
+struct HttpClient : public AsyncEventSource {
```

Comment:
HttpClient is meant to be added to a MessageLoop (pending the asio version). In this case, it's required by the interface of MessageLoop::addSource.